Enable more Ruff lints & fix issues

akx opened the pull request:
# What does this PR do?

Follows up on #2400. 
Sibling of https://github.com/huggingface/trl/pull/1300 and https://github.com/huggingface/peft/pull/1421

I didn't enable the full `B` family yet, because there's still about 90 errors left if I do, and best keep these PRs somewhat reviewable...

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr), Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [ ] Did you write any new necessary tests?

## Who can review?

cc @muellerzr

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/accelerate/pr_2419). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

BenjaminBossan commented on pyproject.toml:
```diff
@@ -1,15 +1,30 @@
+
+[tool.ruff.lint.per-file-ignores]
+"__init__.py" = [
+    "F401",  # Ignore seemingly unused imports (they're meant for re-export)
```
This could be handled by defining `__all__` in the corresponding modules, but I'm fine with simply ignoring those imports.

akx replied:
I had a commit for that in here too, but it added +300 lines, so I decided to skip it. Can easily add in a subsequent PR.

BenjaminBossan reviewed (changes requested):
Thanks for extending the use of ruff in this repo. The changes look good, I just found one instance which I think is better fixed manually. Please take a look.

akx commented on src/accelerate/utils/megatron_lm.py:
```diff
@@ -845,8 +844,8 @@ def initialize(accelerator, extra_args_provider=None, args_defaults={}):
                 print(
-                    "WARNING: overriding default arguments for {key}:{v} \
-                        with {key}:{v2}".format(key=key, v=getattr(args, key), v2=value),
+                    f"WARNING: overriding default arguments for {key}:{getattr(args, key)} \
```
Yeah, the autofix didn't do anything about the `\` continuation. Will reformat!

akx commented:
Gentle ping @BenjaminBossan :)

BenjaminBossan reviewed (approved):
Thanks for the adjustments, LGTM. Leaving it up to @muellerzr to merge.

And please always ping when the PR is ready for the next review round.

BenjaminBossan commented on src/accelerate/utils/megatron_lm.py:
```diff
@@ -845,8 +844,8 @@ def initialize(accelerator, extra_args_provider=None, args_defaults={}):
                 print(
-                    "WARNING: overriding default arguments for {key}:{v} \
-                        with {key}:{v2}".format(key=key, v=getattr(args, key), v2=value),
+                    f"WARNING: overriding default arguments for "
```
the `f` on this line is not needed

muellerzr reviewed (approved):
Thanks for bringing us to 2.x.x for ruff! Can you run `make style; make quality` once more? This should fix the last failure case (in `megatron`)

akx commented:
> Can you run `make style; make quality` once more? 

Rebased (and fixed the couple new `UP` complaints that popped up from `pippy` examples), done.

BenjaminBossan commented:
I did a re-check after the latest changes, still LGTM.
